### ISSUE TITLE
Phase 2: Triton コード生成・ベンチマーク・subprocess ワーカー

### ISSUE THREAD

**Issue by flipslidersand** [flipslidersand-labs/forge#2]
## 概要

Phase 1（IR + キャッシュ基盤）完了済み。Phase 2 では GPU が必要なコンポーネントを実装し、
RMSNorm の Triton カーネルを生成→コンパイル→ベンチマークまで動かす。

**前提**: GPU 環境が必要。先に #3 の調査を完了させること。

---

### 1. Triton コード生成

- [ ] `src/forge/codegen/triton_codegen.py` — `generate_rmsnorm(spec, params) -> str`
- [ ] `src/forge/codegen/templates/rmsnorm.py.jinja` — Jinja2 テンプレート
- [ ] `tests/test_codegen.py` — 生成コードが `compile()` を通ること（GPU なしで検証可）

パラメータ:

```python
@dataclass(frozen=True)
class SearchParams:
    block_size: int    # 512 | 1024 | 2048 | 4096
    num_warps: int     # 4 | 8 | 16
    num_stages: int    # 1 | 2 | 3 | 4
    acc_dtype: str     # "fp32" | "fp16"
    variant: str       # "single_row"（Phase 2 では固定）
```

### 2. subprocess ワーカー

- [ ] `src/forge/runtime/_worker_entry.py` — stdin から JSON を受け取り実行・結果を stdout へ
- [ ] `src/forge/runtime/worker.py` — `run_in_worker(code, inputs, task, timeout_s) -> dict`
- [ ] `src/forge/runtime/launcher.py` — ワーカー管理（タイムアウト・エラーハンドリング）

入出力プロトコル:

```json
// stdin
{"code": "...", "inputs": [{"shape": [2048, 4096], "dtype": "float16", "seed": 42}], "task": "benchmark"}

// stdout
{"success": true, "median_us": 42.3, "p20_us": 40.1, "p80_us": 44.7, "samples": [...]}
// または
{"success": false, "error": "CUDA illegal memory access"}
```

### 3. GPU ベンチマーク

- [ ] `src/forge/benchmark/timer.py` — `measure(fn, *args, warmup=25, repeat=200) -> BenchmarkResult`
- [ ] `src/forge/benchmark/statistics.py` — `is_improvement(candidate, baseline, min_speedup=1.03)`
- [ ] `tests/test_benchmark.py`

```python
# is_improvement の判定ロジック
def is_improvement(candidate, baseline, min_speedup=1.03):
    return candidate.p80_us < baseline.p20_us / min_speedup
```

### 4. 動作確認

```bash
python -c "
from forge.codegen.triton_codegen import generate_rmsnorm
# ... コード生成 → subprocess で実行 → BenchmarkResult を表示
"
```

---

## 完成条件

- [ ] 生成した Triton コードが GPU で実行できる
- [ ] `BenchmarkResult(median_us, p20_us, p80_us)` が返る
- [ ] CUDA illegal memory access が起きても親プロセスが死なない
- [ ] `tests/test_codegen.py` 全通過（GPU なし環境でも通ること）

## 依存

- #3 (GPU 環境・Triton バージョン調査) が先決

**Comment by flipslidersand** [comment]
## ✅ Phase 2 実装完了 (commit e1cee41)

### 完成したもの

| コンポーネント | ファイル | 状態 |
|---|---|---|
| Triton codegen | `codegen/triton_codegen.py` + `templates/rmsnorm.py.jinja` | ✅ |
| SearchParams | `search/params.py`（power-of-2 / variant 検証付き） | ✅ |
| 参照実装 | `runtime/reference.py`（fp32 accumulate） | ✅ |
| ベンチマーク | `benchmark/timer.py`（CUDA Event）+ `statistics.py` | ✅ |
| subprocess worker | `runtime/worker.py` + `_worker_entry.py` | ✅ |

### テスト: 44 passed
- CPU 40（codegen/benchmark/ir/cache）
- GPU 4（`@pytest.mark.gpu`、CUDA 無し環境では skip）

### GPU end-to-end 検証（GTX 1080, cc 6.1）
- 正常候補 (block=4096): correct=True, **max_diff=1.95e-3**, benchmark 取得 ✅
- 不正候補 (block=2048 < N=4096): correct=**False** で弾かれ、**親プロセスは生存** ✅
- timeout / crash も `WorkerResult(success=False)` で安全に回収 ✅

### コード生成戦略の決定（Issue #6 関連）
**Jinja2 → 完全な standalone .py モジュール**を採用。理由:
- `@triton.jit` がソースをファイルから読むため worker は .py 必須（Issue #3 申し送り）
- 生成物がそのまま動く完全なコード → Phase 5 の LLM 編集も容易
→ #6 はこの方針で実質クローズ可。`ADR-006` 化は任意。

---

## ⚠️ Phase 3 への申し送り（baseline の公平性）

現在の baseline は `reference.py` の **fp32-upcast 素朴実装**（遅い）。
speedup が 9x と出るが、これは過大評価。Phase 3 では以下も baseline に追加すること:
- [ ] `torch.nn.functional.rms_norm`（最適化済み組み込み）
- [ ] `torch.compile` 版
- [ ] `@triton.autotune` 版

`num_stages` は Pascal で効果がない可能性 → Phase 3 の SearchSpace で cc によりフィルタする。

**Comment by flipslidersand** [comment]
## ✅ Phase 3 完了 (commit 51b4081)

end-to-end の自動探索パイプラインが動作。

### 実装
- `search/space.py` `SearchSpace.enumerate()` — block_size>=N 制約 + Pascal(cc<7.0) で num_stages=[1] フィルタ
- `search/grid.py` `GridSearch` — 全列挙（差し替え可能な探索器の最初の実装）
- `validation/tolerance.py` + `test_cases.py` — atol=2e-3 + 7 エッジケース
- `runtime/` — worker を複数ケース正確性検証に拡張、**速度 baseline を `F.rms_norm` に変更**（Phase 2 申し送り対応）
- `orchestrator.py` — 探索→検証→ベンチ→キャッシュ

### テスト: 64 passed（CPU 61 + GPU 3）

### GTX 1080 実走結果
```
12 候補 → fp32 acc 6 件 correct / fp16 acc 6 件 INCORRECT で棄却
best: block=4096/warps=4/fp32 = 167.9us
baseline F.rms_norm = 1428us → 8.5x
2 回目: cache HIT 1.6s（探索ゼロ）
```

### Phase 4 への申し送り
- baseline に `torch.compile` / `@triton.autotune` 追加（より厳しい競合）
- variant 軸（multi_row / two_pass / fused_weight）を SearchSpace に追加
- `random_search.py`（grid が大きくなるため）
